lib/src/pubspec_utils.dart

```diff
@@ -37,6 +38,63 @@ Pubspec stripDependencyOverrides(Pubspec original) {
   );
 }
 
+Future<Pubspec> constrainedToAtLeastNullSafetyPubspec(
+    Pubspec original, SystemCache cache) async {
+  /// Get the first version of [package] opting in to null-safety.
+  Future<VersionRange> constrainToFirstWithNullSafety(
+      PackageRange packageRange) async {
+    final ref = packageRange.toRef();
+    final available = await cache.source(ref.source).getVersions(ref);
+    if (available.isEmpty) {
+      return stripUpperBound(packageRange.constraint);
+    }
+
+    available.sort((x, y) => x.version.compareTo(y.version));
+
+    for (final p in available) {
+      final pubspec = await cache.source(ref.source).describe(p);
+      if (pubspec.languageVersion.supportsNullSafety) {
+        return VersionRange(min: p.version, includeMin: true);
+      }
+    }
+    return stripUpperBound(packageRange.constraint);
+  }
+
+  Future<List<PackageRange>> allConstrainedToAtLeasNullSafety(
+    Map<String, PackageRange> constrained,
+  ) async {
+    final result = <PackageRange>[];
+
+    for (final name in constrained.keys) {
+      final packageRange = constrained[name];
+      var unconstrainedRange = packageRange;
+
+      /// We only need to remove the upper bound if it is a hosted package.
+      if (packageRange.source is HostedSource) {
+        unconstrainedRange = PackageRange(
+            packageRange.name,
+            packageRange.source,
+            await constrainToFirstWithNullSafety(packageRange),
+            packageRange.description,
+            features: packageRange.features);
+      }
+      result.add(unconstrainedRange);
```

Comment:
Consider making this parallel, otherwise it might be somewhat slow..

Reply:
Good point
Done